tap_salesforce/client.py

```diff
@@ -188,25 +179,88 @@ def get_records(
             limit_stm = f"LIMIT {limit}"
         else:
             limit_stm = ""
+        query = f"{select_stm} {from_stm} {where_stm} {order_by_stm} {limit_stm}"
+        return query
+
+    def field_chunker(
+        self, fields: List[str], size: int
+    ) -> Generator[List[str], None, None]:
+        field_chunk = []
+        length = 0
+        index = 0
+        for field in fields:
+            index += 1
+            length += len(field)
+            field_chunk.append(field)
+            if (length > size) or (index == len(fields)):
+                yield field_chunk
+                field_chunk = []
+                length = 0
+
+    def merge_records(
+        self, paginators: List[Generator[Dict, None, None]], table: Table
+    ) -> Generator[Dict, None, None]:
+        for records in zip(*paginators):
+            merged_record = {}
+            primary_key = None
+            for record in records:
+                if not primary_key:
+                    primary_key = record[table.primary_key]
+                if primary_key != record[table.primary_key]:
+                    raise PrimaryKeyNotMatch(
+                        f"couldn't merge records with different primary keys: {primary_key} and {record[table.primary_key]}"
+                    )
+                merged_record.update(record)
 
-        LOGGER.info(
-            f"""
-            {select_stm}
-            {from_stm}
-            {where_stm}
-            {order_by_stm}
-            {limit_stm}
-        """
-        )
+            yield merged_record
 
-        query = f"{select_stm}{from_stm}{where_stm}{order_by_stm}{limit_stm}"
+    def get_records(
+        self,
+        table: Table,
+        fields: List[str],
+        start_date: datetime,
+        end_date: Optional[datetime] = None,
+        limit: Optional[int] = None,
+        shrink_window_factor: int = 2,
+    ):
 
+        query = self.construct_query(table, fields, start_date, end_date, limit)
         try:
-            yield from self._paginate(
-                "GET",
-                f"/services/data/{self._API_VERSION}/queryAll/",
-                params={"q": query},
-            )
+            if len(query) <= MAX_QUERY_LENGTH:
+                LOGGER.info(query)
+                yield from self._paginate(
+                    "GET",
+                    f"/services/data/{self._API_VERSION}/queryAll/",
+                    params={"q": query},
+                )
+            elif table.primary_key:
+                LOGGER.info(f"query too long {len(query)=}, split into subqueries")
+                paginators = []
+                for field_chunk in self.field_chunker(fields, 8000):
+                    field_chunk.append(table.primary_key)
+                    field_chunk.append(table.replication_key)
```

Comment:
Wouldn't this cause duplicate fields, since a table likely has its primary- and replication keys in its list of fields already?

Reply:
If we split the query, the second query won't have primary key and replication_key. I handle deduplication in `construct_query` function